src/backends/perf_pt/decode.c

```diff
@@ -35,13 +35,317 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
+#define _GNU_SOURCE
+
+#include <stdio.h>
 #include <intel-pt.h>
+#include <pt_cpu.h>
+#include <stdbool.h>
+#include <inttypes.h>
+#include <stdint.h>
+#include <link.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
 
-void
-perf_pt_start_decode(void *buf) {
-    // XXX This function body is only here to check for correct libipt linkage
-    // at this time.
-    (void) buf;
+#include "perf_pt_private.h"
+
+struct load_self_image_args {
+    struct pt_image *image;
+    int vdso_fd;
+    char *vdso_filename;
+};
+
+// Private prototypes.
+static int drain_events(struct pt_block_decoder *, int);
+static bool load_self_image(struct load_self_image_args *);
+static int load_self_image_cb(struct dl_phdr_info *, size_t, void *);
+
+// Public prototypes.
+void *perf_pt_start_decode(void *, uint64_t, int, char *, int *);
+bool perf_pt_next_block(struct pt_block_decoder *, int *, uint64_t *);
+void perf_pt_free_decoder(struct pt_block_decoder *);
+
+/*
+ * Start decoding a PT trace.
+ *
+ * Accepts a raw buffer `buf` of length `len`.
+ *
+ * `vdso_fd` is an open file descriptor for the filename `vdso_filename`. This
+ * is where the VDSO code will be written. libipt will read this file lazily,
+ * so it's up to the caller to make sure this file lives long enough for their
+ * purposes.
+ *
+ * `*decoder_status` will be updated to reflect the status of the decoder after
+ * it has been synchronised.
+ *
+ * Returns a pointer to a configured libipt block decoder or NULL on error.
+ *
+ * XXX This function assumes that we are decoding a trace for the current CPU
```

Comment:
This doesn't feel like an XXX to me: it's a fundamental assumption of the whole library, right?

Reply:
The CPU part is should probably be baked in.

With regards to "the current process", I don't see why our library couldn't be used more generally to trace other processes. We even offer a configuration option to choose an arbitrary `tid` to trace. But if we wanted to decode a trace for another process, we would have to offer an API to load the code for a process other than ourself.

Reply:
Maybe someone might do that as a PR, but I can't see that we'd ever want this?

Reply:
Here's the question: Is hwtracer a library for our meta-tracer, or is it general purpose?

Reply:
I am not generally a fan of putting in functionality for use cases that we don't know will ever exist. So I think, until someone gives us a use case that suggests that we should generalise, we might as well keep things simple.

Reply:
Fair enough. FWIW, we have #49 to remind us that this is on the cards.

Reply:
I'm not sure having that as an issue is very useful (*we're* never going to do it AFAICS), but up to you.

Reply:
I think the correct thing to do is to update the README and clearly state that the library is currently for tracing one's self.

Reply:
I agree.